dashboard/app/helpers/lti_helper.rb

```diff
@@ -0,0 +1,123 @@
+module LtiHelper
+  #
+  # LTI role can be blank, one, or multiple (comma separated).
+  # https://www.imsglobal.org/specs/ltiv1p0/implementation-guide#toc-19
+  #
+  # e.g. "Instructor, TeachingAssistant, ContentDeveloper"
+  #
+  def lti_role_to_user_type(roles)
+    return User::TYPE_STUDENT if roles.blank?
+
+    arr = roles.downcase.split(/\s*,\s*/).uniq.compact
+    if arr.length == 1 && arr.include?('learner')
+      User::TYPE_STUDENT
+    else
+      User::TYPE_TEACHER
+    end
+  end
+
+  def lti_param_missing_msg(pp)
+    "Sorry, could not sign up--[#{pp}] missing from LTI request. Please contact your LMS admin."
+  end
+
+  #
+  # ===========================================================
+  #   LTI Signin
+  # ===========================================================
+  #
+  def lti_signin_and_redirect(params, redirect_url)
+    if params[:oauth_consumer_key].blank?
+      redirect_to new_user_registration_path,
+        alert: "Sorry, launch request must specify :oauth_consumer_key"
+      return
+    end
+
+    consumer_key = params[:oauth_consumer_key]
+    lti_credentials = CDO.lti_credentials || {}
+    consumer_secret = lti_credentials[consumer_key]
+    if consumer_secret.blank?
+      redirect_to new_user_registration_path,
+        alert: "Sorry, invalid consumer_key [#{consumer_key}]"
+      return
+    end
+
+    provider = IMS::LTI::ToolProvider.new(consumer_key, consumer_secret, params)
+    if provider.valid_request?(request)
+      if params[:oauth_timestamp].blank?
+        redirect_to new_user_registration_path,
+          alert: "Sorry, launch request must specify :oauth_timestamp"
+        return
+      end
+
+      delta = Time.now.to_i - params[:oauth_timestamp].to_i
```

Comment:
Minor: Rails best practice is to use `Time.zone.now` instead of `Time.now`. See http://api.rubyonrails.org/classes/ActiveSupport/TimeWithZone.html. Our production servers however are set to utc so they are equivalent.